src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,17 +105,131 @@ def pileupSizeTask(self):
             self.logger.exception(msg)
         return cmsDict, cusDict
 
+    def partialPileupTask(self):
+        """
+        Execute partial pileup placement according to the following logic:
+
+        - get rucio DIDs (datasets, in CMS name blocks) for our pileup document
+        - get portion of DIDs based on ceil(containerFraction * num_rucio_datasets)
+        - create new container DID in Rucio via Rucio wrapper createContainer API
+          which calls rucio CLI add_container
+        - update MSPileup JSON with
+          - create new custom Name as pileup+extention
+          - add new transition record and update MSPileup document
+        - we call attachDIDs Rucio wrapper API with our set of DIDs and rses from pileup document
+        - create new rules for custom DID via provided rseExpression
+        - set expiration date (to be 24h) for already existing ruleIds from pileup document
+        - add new ruleIds to pileup document
+
+        :return: None
+        """
+        spec = {'active': True}
+        docs = self.mgr.getPileup(spec)
+        for doc in docs:
+            self.logger.info("partialPileupTask process %s", doc)
+            fraction = doc['containerFraction']
+
+            # check previous fraction value and proceed only if has changed
+            previousFraction = 1.0
+            transition = doc.get('transition', [])
+            if transition:
+                previousFraction = transition[-1]['containerFraction']
+            if previousFraction == fraction:
+                # do nothing since there is no transition, i.e. skip this record
+                self.logger.info("Pileup name %s has no container fractions changes", doc['pileupName'])
+                continue
+
+            # get rucio DIDs (datasets, in CMS name blocks) for our pileup document as following:
+            # - when container fraction is decreasing: new container is a subset of customName
+            # - when container fraction is increasing: new container is a superset of customName + subset of pileupName
+
+            # usage of block names defined in this logic:
+            # https://github.com/dmwm/WMCore/pull/11807#pullrequestreview-1786778783
+            pname = doc['pileupName']
+            blockNames = self.rucioClient.getBlocksInContainer(pname)
+            if previousFraction < fraction:
+                # if it increasing
+                # use EVERY single block defined in the custom dataset plus blocks from the standard pileup
+                cname = doc.get('customName', '')
+                if cname:
+                    blockNames = self.rucioClient.getBlocksInContainer(cname) + blockNames
```

Comment:
This is wrong, you will be double counting blocks from the pileup and from the custom pileup.
Blocks from the original pileup must be used to define how many blocks must be attached to the new custom pileup. In other words, this gives us the baseline of how many blocks must be present in the new container. This value is then used in the below logic (nothing new here, I am just trying to explain this functionality with different words).

In addition:
a) if container is increasing: use every single block from the previous custom container, plus the remaining from the original pileup (randomly)
b) if container is decreasing: simply use blocks from the custom container (randomly).

Reply:
I'm sorry Alan, I implemented exactly how logic is written, i.e.:
```
# the logic
# a) if container is increasing: use every single block from the previous custom container, plus the remaining from the original pileup (randomly)

blockNames = self.rucioClient.getBlocksInContainer(pname)
if previousFraction < fraction: # if if increasing
   cname = doc.get('customName', '')
   if cname:
      blockNames = self.rucioClient.getBlocksInContainer(cname) + blockNames
```
So, let's break it down:
- I got all blocks from pileup
- if we have increase I get customName and if it exists I get blocks from it and add to them blocks from original pileup.

What exactly is wrong? The double counting will ONLY occur if I'll do `blockNames += self.rucioClient.getBlocksInContainer(cname) + blockNames` , instead I take blocks from customname (previous custom container) and add to them original blocks from pileup. To avoid further discussion, please provide how you'll implement the logic (a) and (b)

Reply:
Valentin, I think the "remaining" word was ambiguous in my explanation, sorry. By remaining, I meant the remaining fraction of blocks, not the remaining blocks from the original dataset.

I updated the gist to reflect details of the container increase/decrease algorithm:
https://gist.github.com/amaltaro/b4f9bafc0b58c10092a0735c635538b5#logic-for-increasingdecreasing-container-fraction

However, if there is anything not yet clear, I am happy to go through these over Zoom.